### PR TITLE
feat: add callbacks for chains and agents

### PR DESCRIPTION
**Pull request by sullivan-sean**

### CODE REVIEW

File: langchain/callbacks/index.ts [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+  setHandlers: (handlers: CallbackHandler[]) => void;
+}
+
+const isEventType = <T extends string>(
+  prefix: T,
+  event: CallbackEvent
+): event is CallbackEvent & { event: `${T}.${string}` } =>
+  event.event.startsWith(prefix);
+
+export abstract class EventTypeCallbackHandler extends BaseCallbackHandler {
```
Comment (nfcampos) [inline langchain/callbacks/index.ts:143]: I think we should make all handlers async (and by extension the manager too)

Comment (sullivan-sean) [reply langchain/callbacks/index.ts:143]: I think in most cases they are async already? (returning `Promise<void> | void`)

Looks like i missed it on the `PerEventCallbackHandler` object but should be easy fix

File: langchain/llms/base.ts [hunk 2]
```diff
@@ -33,15 +26,26 @@ export abstract class BaseLLM {
...
    */
   verbose?: boolean = false;
 
-  constructor(callbackManager?: LLMCallbackManager, verbose?: boolean) {
-    this.callbackManager = callbackManager ?? getCallbackManager();
+  constructor(
+    callbackManager?: LLMCallbackManager | BaseCallbackManager,
+    verbose?: boolean
+  ) {
+    if (!callbackManager) {
```
Comment (hwchase17) [inline langchain/llms/base.ts:40]: some comment here explaining this would probably be nice

**Review by hwchase17 — COMMENTED** [review]
high level:

- we recently updated the langchain code to have a concept of a tool class (which this package does already) and then we moved the callbacks from the agent executor to the tool class (added callbacks before/after tool.call, similar to how we do for chains)
-

File: langchain/agents/executor.ts [hunk 3]
```diff
@@ -74,9 +84,33 @@ export class AgentExecutor extends BaseChain {
...
+          : `${action.tool} is not a valid tool, try another one.`;
+      } catch (err) {
+        this.callbackManager(
+          { event: "agent.tool_error", err },
+          { verbose: this.verbose }
+        );
+        throw err;
+      }
+
+      this.callbackManager(
```
Comment (nfcampos) [inline langchain/agents/executor.ts:109]: With async I meant even this method should be async, otherwise we end up dangling promises

**Review by agola11 — COMMENTED** [review]
@sullivan-sean -- since this PR is a bit stale with lots of merge conflicts, I'm going to take pieces of this and apply them in a new PR, might make some changes here and there. Hope that's okay with you. Will add you as a collaborator on the commit

**Pull request closed without merging**